### DISCUSSION TITLE
6.1.1 behaviour change: no longer possible to reassign functions in observables

### DISCUSSION THREAD

**Discussion by yl92** [mobxjs/mobx#2760] [Q&A] [answered]
---

Update: here is a repo with runnable code for quicker inspection:
https://github.com/troydraws/playground-mobx

---

I had been on MobX 6.0.4 before; after updating to MobX 6.1.1, In the following example, reassigning a new function to the `normalFunc` property will throw, but not `actionFunc`:

```
import { observable, action, isAction } from 'mobx';

const obj = observable({
  value: 1,
  actionFunc: action(() => obj.value++),
  normalFunc: () => console.log('this is a func'),
})

console.log(isAction(obj.actionFunc)); // <- true
console.log(isAction(obj.normalFunc)); // <- true

obj.actionFunc = () => {
  console.log('this is a new actionFunc');
  obj.value++
} // <- runs fine with or without wrapping into action()

obj.normalFunc = () => {
  console.log('this is a new normalFunc')
} // <- TypeError: 'set' on proxy: trap returned falsish for property 'normalFunc'

```

Previously I had been thinking of functions in observables like `normalFunc` and `actionFunc` as observable references to functions. I treat them just like other keys with primitive values and can be overridden any time and trigger updates, which I had been doing extensively... Went to check the 6.1 changelog, and found out that "All annotated and non-observable props (action/flow) are non-writable. Previously writable.". I guess that's why this is now throwing? 

Not sure if this was an intended behaviour as a result of the change?

**Comment by danielkcz** [comment]
If you think you have found a bug, creating an issue is generally better than a discussion. Either way, can you provide runnable reproduction that we can inspect?

**Reply by yl92** [reply]
Thanks Fredy, I wasn't sure if it is a bug or if it was intended behaviour so thought it'd be more fitting to post a question instead. The code snippet inside the question is directly runnable :)

**Reply by yl92** [reply]
https://github.com/troydraws/playground-mobx <- So i made one with a bit more details & experiments, have a play! It all works fine with 6.0.4.

**Comment by mweststrate** [comment, marked as answer]
But yes the behaviour has been made explicit. So what you might be looking
for to have observable actions is: annotate the field with observable, and
wrap the value in action.

**Reply by yl92** [reply]
Thanks Michel, that sounds like quite a bit of extra work... haha for the time being I might just stick with an older version. I've updated with a little demo repo to show the differences if you are interested to see, though I guess you'd be expecting to see the results anyways: https://github.com/troydraws/playground-mobx

**Comment by urugator** [comment]
Since mobx 6.0 all functions are auto converted to `action` (unless you specify otherwise). Action fields were never observable.
Since mobx 6.1 these non-observable fields are not writable to prevent errors where you would assume that changing the value (function) would notify observers.
If the function is already an `action`, we simply ignore the field, sort of assuming it's already set-up properly... However I agree that may not be an ideal behavior as you end up in a similar situation which we want to prevent by making actions non-writable ... not entirely sure about this...
We are also considering that auto converted functions become both observable and actions... (therefore writable and additionally enumerable)

**Reply by urugator** [reply]
For reference
https://github.com/mobxjs/mobx/blob/28f8a11d8b94f1aca2eec4ae9c5f45c5ea2f4362/packages/mobx/src/api/annotation.ts#L37-L66
I think we could change it (at least for the time being), so that if already action/flow return `observable`, but I am not entirely sure atm if we can do that without breaking something else and whether it makes sense to do so.

**Reply by yl92** [reply]
Thanks urugator, Interesting to know that they are designed to be non-observable - i've been using MobX with React in this way for a while (since before 6 came about), haven't noticed replacing actions not triggering updates! Bit curious why the decision was made (was it explained in docs somewhere that I didn't see?), since I suppose if I treat flow/actions as just another type of "things" to be passed around, it feels natural to think of them as replaceable and observable just like everything else.

**Reply by yl92** [reply]
https://github.com/troydraws/playground-mobx <- So i made one with a bit more details & experiments, have a play! It all works fine with 6.0.4.

**Reply by urugator** [reply]
Well we try to do what seems the most practical and typically functions/methods represent logic that is static - doesn't change over time.
There was always a demand for functions being automatically actions, because it would make it less tedious to define simple stores (like local component state) and it's something you want most of the time anyway.
The observability of such fields wasn't actually much of a concern - the problem was that we couldn't distinguish between functions that are supposed to update state and helper functions that are used as part of derivations.
However in V6 we came up with `autoAction` that can be used in both cases - so "finally" we were able to do what we/users wanted...
There are obviously different opinions about this, you may want to check these cca 3 comments for some additional reasoning:
https://github.com/mobxjs/mobx/pull/2641#issuecomment-737206607

**Reply by urugator** [reply]
Can you try to print 
`isAction(o.key)`
`isObservableProp(o, key)`
for each function, before and after the assigment?

**Reply by yl92** [reply]
thanks, that thread does give a bit more insight, i'll have a think/play later on... I've added the isAction & isObservableProp logs now.

**Reply by yl92** [reply]
![image](https://user-images.githubusercontent.com/13817751/106209895-c5317c80-61bd-11eb-99fa-bd6394f95887.png)
^^ logs with 6.0.4

**Reply by yl92** [reply]
![image](https://user-images.githubusercontent.com/13817751/106209983-e85c2c00-61bd-11eb-862a-8708cba54971.png)
^^ 6.1.1

**Reply by urugator** [reply]
As you can see there is a "bug" in 6.0 ... firstly the observer somehow subscribes for prop that is not observable and then the property suddenly becomes observable on assigment. What would be the point of making field non-observable if it can't prevent subscription and becomes observable on assigment...
Also you would probably get different behavior with `proxy: false` option (on 6.0).

**Reply by yl92** [reply]
I saw that yeah, a little funny 😅 Let me have a play this weekend to figure things out a bit more... thanks for the help!
